mam/src/mam.rs

```diff
@@ -4,27 +4,25 @@ use sign::iss;
 use merkle;
 use trytes::*;
 use tmath::*;
-use auth::*;
 use mask::*;
 use errors::*;
+use pascal;
 
 /// generate the message key for a root and an index.
 /// It copies `root` to `key`, and adds `index` to it.
-pub fn message_key(root: &[Trit], index: usize, key: &mut [Trit]) -> Result<usize, MamError> {
-    if key.len() < root.len() {
-        Err(MamError::ArrayOutOfBounds)
-    } else {
-        key.clone_from_slice(root);
-        add_assign(&mut key[..HASH_LENGTH], index as isize);
-        Ok(key.len())
-    }
+pub fn message_key<C: Curl<Trit>>(side_key: &[Trit], root: &[Trit], index: usize, curl: &mut C) {
+    curl.absorb(side_key);
+    let mut root_copy: [Trit; HASH_LENGTH] = [0; HASH_LENGTH];
+    root_copy.clone_from_slice(root);
+    add_assign(&mut root_copy, index as isize);
+    curl.absorb(&root_copy);
 }
 
 /// generates the address for a given mam `key`
 /// for a mam, the key should consist of the merkle root and
 /// an initialization vector, which is the index of the key  in the
 /// merkle tree being used
-pub fn message_id<T, C>(key: &mut [T], curl: &mut C)
+pub fn message_id<T, C>(curl: &mut C, key: &mut [T])
```

Comment:
Why the reordering?
I guess you expect people to provide the same curl as for `message_key` here? 
If not, then above comment holds because you explicitly require a key variable here.

Also, documentation update?

Reply:
No, you're right. I shouldn't have changed this part.